cql3/expr/prepare_expr.cc

```diff
@@ -1239,6 +1239,20 @@ binary_operator prepare_binary_operator(binary_operator binop, data_dictionary::
     lw_shared_ptr<column_specification> rhs_receiver = get_rhs_receiver(lhs_receiver, binop.op);
     expression prepared_rhs = prepare_expression(binop.rhs, db, table_schema.ks_name(), &table_schema, rhs_receiver);
 
+    // IS NOT NULL requires an additional check that the RHS is NULL.
+    // Otherwise things like `int_col IS NOT 123` would be allowed - the types match, but the value is wrong.
+    if (binop.op == oper_t::IS_NOT) {
+        bool rhs_is_null = is<constant>(prepared_rhs) && as<constant>(prepared_rhs).is_null();
+
+        if (!rhs_is_null) {
+            expression binop_expr = binop;
+            expression::printer binop_printer{.expr_to_print = binop_expr, .debug_mode = false};
+            throw exceptions::invalid_request_exception(format(
```

Comment:
We should have a with_debug(const expression&) helper to reduce boilerplate here.